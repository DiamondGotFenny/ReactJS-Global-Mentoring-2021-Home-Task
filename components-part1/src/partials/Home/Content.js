import React, { useState, memo, useEffect } from 'react';
import { useNavigate, useParams, useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';
import searchObject from 'search-object';
import queryString from 'query-string';
import MovieCard from '../../components/MovieCard';
import {
  SectionWrapper,
  FilterButton,
  FilterWrapper,
  SortSelect,
  SortSpan,
  SortWrapper,
  StyledResults,
  ContentWrapper,
} from '../../styledComponents/Content';
import httpService from '../../services/httpService';

const Content = () => {
  const params = useParams();
  const queryStr = useLocation().search;
  const queryObj = queryString.parse(queryStr);
  const { pathname } = useLocation();
  const navigate = useNavigate();
  const deletedMovie = useSelector((state) => state.deletedMovie);
  console.log(deletedMovie, 'deletedMovie');
  const [moviesList, setmoviesList] = useState([]);
  //this is the original list of movies
  const [movies, setMovies] = useState(null);

  const handleSortMovies = (movies, sortBy) => {
    if (sortBy === 'vote_average') {
      const sortedMovies = movies.sort(
        (a, b) => b.vote_average - a.vote_average
      );
      return sortedMovies;
    } else if (sortBy === 'release_date') {
      const sortedMovies = movies.sort(
        (a, b) => Date.parse(b.release_date) - Date.parse(a.release_date)
      );
      return sortedMovies;
    }
  };

  useEffect(() => {
    const fetchData = async () => {
      try {
        const { data } = await httpService.get('/movies');
        setMovies(data);
        setmoviesList(data.data);
      } catch (error) {
        alert(error);
      }
    };
    fetchData();
  }, []);
  console.log(movies, 'movies');
  useEffect(() => {
    if (queryObj.sortBy) {
      const sortedMovies = handleSortMovies(moviesList, queryObj.sortBy);
      setmoviesList(sortedMovies);
    }
  }, [queryObj.sortBy]);

  useEffect(() => {
    if (movies && !queryObj.searchQuery) {
      setmoviesList(movies.data);
    }
    if (movies && params.searchQuery) {
      setmoviesList(handleSearchQuery(movies.data, params.searchQuery));
    }
  }, [movies, params.searchQuery]);

  useEffect(() => {
    if (movies && queryObj.genre) {
      setmoviesList(handleSearchQuery(movies.data, queryObj.genre));
    }
  }, [movies, queryObj.genre]);

  const handleSearchQuery = (movies, query) => {
    const filteredMovies = movies
      .map((movie) => {
        if (searchObject(movie, query)) {
          return movie;
        }
      })
      .filter((item) => item !== undefined);
    return filteredMovies;
  };

  const handleGenreClick = (genre) => {
    if (genre === 'all') {
      setmoviesList(movies.data);
      navigate(`${pathname}`);
      return;
    }

    const newQueryObj = { ...queryObj, genre: genre };
    const newQueryStr = queryString.stringify(newQueryObj);
    navigate({ search: newQueryStr });
  };
  const handleInputChange = (e) => {
    const { value } = e.target;
    const newQueryObj = { ...queryObj, sortBy: value };
    const newQueryStr = queryString.stringify(newQueryObj);
    navigate({ search: newQueryStr });
  };
  return (
    <SectionWrapper>
      <FilterWrapper>
        <ul>
          <li onClick={() => handleGenreClick('all')}>
            <FilterButton>all</FilterButton>
          </li>
          <li onClick={() => handleGenreClick('Drama')}>
            <FilterButton>Drama</FilterButton>
          </li>
          <li onClick={() => handleGenreClick('Comedy')}>
            <FilterButton>Comedy</FilterButton>
          </li>
          <li onClick={() => handleGenreClick('Adventure')}>
            <FilterButton>Adventure</FilterButton>
          </li>
          <li onClick={() => handleGenreClick('Action')}>
            <FilterButton>Action</FilterButton>
          </li>
        </ul>
        <SortWrapper>
          <SortSpan>Sort by</SortSpan>
          <SortSelect onChange={handleInputChange} defaultValue="release_date">
            <option value="release_date">release date</option>
            <option value="vote_average">rating</option>
          </SortSelect>
        </SortWrapper>
      </FilterWrapper>
      <StyledResults>
        <span>{moviesList.length > 0 ? moviesList.length : 0}</span> movies
        found
      </StyledResults>
      <ContentWrapper>
        {moviesList.map((movie) => (
          <MovieCard movie={movie} key={movie.id} />
        ))}
      </ContentWrapper>
    </SectionWrapper>
  );
};

export default memo(Content);
