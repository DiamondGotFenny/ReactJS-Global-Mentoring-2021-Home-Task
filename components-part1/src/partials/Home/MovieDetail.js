import React, { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { Image } from 'react-bootstrap';
import styled from 'styled-components';
import Logo from '../../components/Logo';
import httpService from '../../services/httpService';

const StyledMovieDetailWrapper = styled.div`
  height: 396px;
  width: 1260px;
  display: grid;
  .logo-container {
    grid-area: logo;
    padding-top: 20px;
  }
  .search-btn {
    grid-area: search;
    background-color: transparent;
    border: none;
    outline: none;
    color: #f65261;
  }
  .poster {
    grid-area: poster;
    width: 220px;
    height: auto;
    transform: translateX(25%);
  }
  .info-container {
    grid-area: info;
    width: 780px;
    height: auto;
    padding-top: 10px;
    display: flex;
    flex-direction: column;
    .col-1,
    .col-2 {
      width: 100%;
      display: flex;
    }
    .rating {
      border-radius: 50%;
      width: 40px;
      height: 40px;
      margin-left: 10px;
      border: 1px solid #fff;
      font-size: 18px;
      line-height: 35px;
    }
    .genre {
      text-align: left;
    }
    .runtime {
      margin-left: 10px;
    }
    .release-date,
    .runtime {
      color: #f65261;
    }
    .desc {
      margin-top: 10px;
      text-align: left;
    }
  }
  grid-template-areas:
    'logo . . search'
    'poster poster info info'
    'poster poster info info';
  margin: 0;
  background-color: #232323;
  color: #fff;
`;

const MovieDetail = ({ movieId }) => {
  const navigate = useNavigate();
  const [movieDetails, setMovieDetails] = React.useState(null);

  useEffect(() => {
    const fetchMovieDetails = async () => {
      console.log('movieId', movieId);
      try {
        const { data } = await httpService.get(`/movies/${movieId}`);
        setMovieDetails(data);
      } catch (error) {
        alert(error);
      }
    };
    fetchMovieDetails();
  }, [movieId]);
  if (!movieDetails) {
    return <StyledMovieDetailWrapper>No Found</StyledMovieDetailWrapper>;
  }

  return (
    <StyledMovieDetailWrapper>
      <Logo />
      <button className="search-btn" onClick={() => navigate('/search')}>
        Search
      </button>
      <Image className="poster" src={movieDetails.poster_path} />
      <div className="info-container">
        <div className="col-1">
          <h2 className="title">{`${movieDetails.title}:${movieDetails.tagline}`}</h2>
          <p className="rating">{movieDetails.vote_average}</p>
        </div>

        <p className="genre">{movieDetails.genres.join(' ')}</p>
        <div className="col-2">
          <p className="release-date">{movieDetails.release_date}</p>
          <p className="runtime">{movieDetails.runtime} mints</p>
        </div>

        <p className="desc">{movieDetails.overview}</p>
      </div>
    </StyledMovieDetailWrapper>
  );
};

export default MovieDetail;
