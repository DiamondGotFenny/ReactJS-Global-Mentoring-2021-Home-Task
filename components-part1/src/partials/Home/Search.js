import React from 'react';
import Logo from '../../components/Logo';
import {
  SearchWrapper,
  StyledHeader,
  StyledAddMovieButton,
  StyledSearchTitle,
  StyledSearchContainer,
} from '../../styledComponents/Search';

const Search = (props) => {
  return (
    <SearchWrapper>
      <StyledHeader>
        <Logo />
        <StyledAddMovieButton>+ ADD MOVIE</StyledAddMovieButton>
      </StyledHeader>
      <StyledSearchTitle>FIND YOUR MOVIE</StyledSearchTitle>
      <StyledSearchContainer>
        <input
          type="search"
          id="site-search"
          name="search-movie"
          aria-label="Search movie"
          placeholder="What do you want to watch?"
        />
        <button>Search</button>
      </StyledSearchContainer>
    </SearchWrapper>
  );
};

export default Search;
