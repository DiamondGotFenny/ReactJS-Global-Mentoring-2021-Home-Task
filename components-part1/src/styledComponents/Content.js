import styled from 'styled-components';

export const SectionWrapper = styled.section`
  display: grid;
  background-color: #232323;
  margin-top: 20px;
  padding: 0 40px;
`;

export const FilterWrapper = styled.div`
  color: #fff;
  display: flex;
  justify-content: space-between;
  border-bottom: 1px solid #ccc;

  & ul {
    display: flex;
  }
  & li {
    margin-right: 10px;
    text-transform: uppercase;
  }
`;

export const FilterButton = styled.button`
  padding: 15px 0 9px 0;
  background-color: transparent;
  text-transform: uppercase;
  border: none;
  color: #fff;
  height: 100%;

  &:focus-within {
    border-bottom: 2px solid #f65261;
  }
`;

export const SortWrapper = styled.div`
  padding: 15px 0 9px 0;
`;

export const SortSpan = styled.span`
  text-transform: uppercase;
  margin-right: 20px;
  font-family: Montserrat;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 20px;
  /* identical to box height */

  letter-spacing: 0.888889px;

  color: #ffffff;

  mix-blend-mode: normal;
  opacity: 0.6;
`;
export const SortSelect = styled.select`
  text-transform: uppercase;
  background-color: transparent;
  border: none;
  color: #fff;
  &:active,
  &:focus {
    outline: none;
    box-shadow: none !important;
  }
  & option:not(:checked) {
    border: none;
    box-shadow: none !important;
    outline: none;
    background: #232323;
  }
  & option:checked {
    box-shadow: none !important;
    outline: none;
    background: #232323;
  }

  & option:focus {
    border: none;
    outline: none;
    border: none;
    box-shadow: none;
  }
  //this is the styles for the dropdown arrow
  padding: 10px;
  padding-right: 30px;
  -moz-appearance: none;
  -webkit-appearance: none;
  appearance: none;
  background-image: url('data:image/svg+xml;charset=US-ASCII,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22292.4%22%20height%3D%22292.4%22%3E%3Cpath%20fill%3D%22%23F65261%22%20d%3D%22M287%2069.4a17.6%2017.6%200%200%200-13-5.4H18.4c-5%200-9.3%201.8-12.9%205.4A17.6%2017.6%200%200%200%200%2082.2c0%205%201.8%209.3%205.4%2012.9l128%20127.9c3.6%203.6%207.8%205.4%2012.8%205.4s9.2-1.8%2012.8-5.4L287%2095c3.5-3.5%205.4-7.8%205.4-12.8%200-5-1.9-9.2-5.5-12.8z%22%2F%3E%3C%2Fsvg%3E');
  background-repeat: no-repeat, repeat;
  background-position: right 0.7em top 50%, 0 0;
  background-size: 0.65em auto, 100%;
`;

export const StyledResults = styled.div`
  margin-top: 20px;
  color: #fff;
  font-family: Montserrat;
  font-style: normal;
  font-weight: 400;
  font-size: 20px;
  line-height: 24px;
  text-align: left;
  & span {
    font-weight: 600;
  }
`;

export const ContentWrapper = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-auto-rows: minmax(500px, auto);
  margin: auto -26px;
`;
