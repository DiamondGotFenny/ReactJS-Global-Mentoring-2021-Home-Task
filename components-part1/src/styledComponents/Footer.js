import styled from 'styled-components';

export const FooterWrapper = styled.footer`
  width: 100%;
  height: 70px;
  background: #424242;
  display: flex;
  justify-content: center;
  align-items: center;
`;
