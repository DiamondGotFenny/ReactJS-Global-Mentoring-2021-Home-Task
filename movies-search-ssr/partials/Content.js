import React, { useEffect } from 'react';
import router, { useRouter } from 'next/router';
import searchObject from 'search-object';
import queryString from 'query-string';
import MovieCard from '../components/MovieCard';
import styles from './Content.module.css';

const Content = ({ movies }) => {
  const { query } = useRouter();
  const [moviesList, setmoviesList] = React.useState([]);
  const handleSortMovies = (movies, sortBy) => {
    if (!movies) return;
    if (sortBy === 'vote_average') {
      const sortedMovies = movies.sort(
        (a, b) => b.vote_average - a.vote_average
      );
      console.log(sortedMovies, 'vote_average');
      return sortedMovies;
    } else if (sortBy === 'release_date') {
      const sortedMovies = movies.sort(
        (a, b) => Date.parse(b.release_date) - Date.parse(a.release_date)
      );
      console.log(sortedMovies, 'release_date');
      return sortedMovies;
    }
  };

  const handleSearchQuery = (movies, query) => {
    const filteredMovies = movies
      .map((movie) => {
        if (searchObject(movie, query)) {
          return movie;
        }
      })
      .filter((item) => item !== undefined);
    return filteredMovies;
  };

  const handleGenreClick = (genre) => {
    if (genre === 'all') {
      setmoviesList(movies);
      router.push(`/search`);
      return;
    }

    const newQueryObj = { ...query, genre: genre };
    const newQueryStr = queryString.stringify(newQueryObj);
    router.push({ search: newQueryStr });
  };
  const handleInputChange = (e) => {
    const { value } = e.target;
    const newQueryObj = { ...query, sortBy: value };
    const newQueryStr = queryString.stringify(newQueryObj);
    router.push({ query: newQueryStr });
  };

  useEffect(() => {
    if (movies) setmoviesList(movies);
  }, [movies]);

  useEffect(() => {
    if (query.sortBy) {
      const sortedMovies = handleSortMovies(moviesList, query.sortBy);
      setmoviesList(sortedMovies);
    }
  }, [query.sortBy]);

  useEffect(() => {
    if (movies && !query.searchQuery) {
      setmoviesList(movies);
    }
    if (movies && query.searchQuery) {
      setmoviesList(handleSearchQuery(movies, query.searchQuery));
    }
  }, [movies, query.searchQuery]);

  useEffect(() => {
    if (movies && query.genre) {
      setmoviesList(handleSearchQuery(movies, query.genre));
    }
  }, [movies, query.genre]);

  return (
    <section className={styles.content}>
      <div className={styles.FilterWrapper}>
        <ul>
          <li onClick={() => handleGenreClick('all')}>
            <button className={styles.FilterButton}>all</button>
          </li>
          <li onClick={() => handleGenreClick('Drama')}>
            <button className={styles.FilterButton}>Drama</button>
          </li>
          <li onClick={() => handleGenreClick('Comedy')}>
            <button className={styles.FilterButton}>Comedy</button>
          </li>
          <li onClick={() => handleGenreClick('Adventure')}>
            <button className={styles.FilterButton}>Adventure</button>
          </li>
          <li data-testid="Action" onClick={() => handleGenreClick('Action')}>
            <button className={styles.FilterButton}>Action</button>
          </li>
        </ul>

        <div className={styles.SortWrapper}>
          <span className={styles.SortSpan}>Sort by</span>
          <select
            className={styles.SortSelect}
            onChange={handleInputChange}
            defaultValue="release_date"
          >
            <option value="release_date">release date</option>
            <option value="vote_average">rating</option>
          </select>
        </div>
      </div>
      <div className={styles['moviesList-counts']}>
        <span>{moviesList?.length > 0 ? moviesList.length : 0}</span>{' '}
        {moviesList?.length > 1 ? 'movies' : 'movie'} found
      </div>
      <div className={styles.ContentWrapper}>
        {!moviesList && <h2>Loading</h2>}
        {moviesList &&
          moviesList.map((movie) => <MovieCard movie={movie} key={movie.id} />)}
      </div>
    </section>
  );
};

export default Content;
