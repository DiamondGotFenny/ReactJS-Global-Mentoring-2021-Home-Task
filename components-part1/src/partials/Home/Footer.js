import React from 'react';
import Logo from '../../components/Logo';
import { FooterWrapper } from '../../styledComponents/Footer';

const Footer = () => {
  return (
    <FooterWrapper className="footer">
      <Logo />
    </FooterWrapper>
  );
};
export default Footer;
