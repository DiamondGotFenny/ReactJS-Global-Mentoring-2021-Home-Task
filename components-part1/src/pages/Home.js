import React, { memo } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import Content from '../partials/Home/Content';
import Footer from '../partials/Home/Footer';
import Search from '../partials/Home/Search';
import styled from 'styled-components';
import MovieDetail from '../partials/Home/MovieDetail';

const HomeWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const Home = () => {
  const queryParams = queryString.parse(useLocation().search);

  return (
    <HomeWrapper>
      {queryParams.movie ? (
        <MovieDetail movieId={queryParams.movie} />
      ) : (
        <Search />
      )}

      <Content />
      <Footer />
    </HomeWrapper>
  );
};

export default memo(Home);
