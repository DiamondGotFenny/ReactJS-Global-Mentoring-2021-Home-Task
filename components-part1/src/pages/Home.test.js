import * as ReactRouter from 'react-router';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Home from './Home';
import Content from '../partials/Home/Content';
import Footer from '../partials/Home/Footer';

//we need to overwrite the useLocation function to writable
//so that we can return a mock location object
Object.defineProperty(ReactRouter, 'useLocation', {
  value: jest.fn(),
  configurable: true,
  writable: true,
});

const setup = (props = {}) => {
  const component = shallow(<Home {...props} />);
  return component;
};

describe('Home', () => {
  it('renders Search correctly when url without movieId query', () => {
    const mockLocation_withoutMovieId = {
      pathname: '/',
      state: {},
      key: '',
      search: '',
      hash: '',
    };
    const useLocationSpy = jest
      .spyOn(ReactRouter, 'useLocation')
      .mockReturnValue(mockLocation_withoutMovieId);
    const wrapper = setup();
    expect(useLocationSpy).toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders MovieDetail correctly with movieId query', () => {
    const mockLocation_withMovieId = {
      pathname: '/',
      state: {},
      key: '',
      search: '?movie=1',
      hash: '',
    };
    const useLocationSpy = jest
      .spyOn(ReactRouter, 'useLocation')
      .mockReturnValue(mockLocation_withMovieId);
    const wrapper = setup();
    expect(useLocationSpy).toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('renders Content component', () => {
    const mockLocation_withoutMovieId = {
      pathname: '/',
      state: {},
      key: '',
      search: '',
      hash: '',
    };
    const useLocationSpy = jest
      .spyOn(ReactRouter, 'useLocation')
      .mockReturnValue(mockLocation_withoutMovieId);
    const wrapper = setup();
    expect(wrapper.containsMatchingElement(<Content />)).toEqual(true);
  });
  it('renders Footer component', () => {
    const mockLocation_withoutMovieId = {
      pathname: '/',
      state: {},
      key: '',
      search: '',
      hash: '',
    };
    const useLocationSpy = jest
      .spyOn(ReactRouter, 'useLocation')
      .mockReturnValue(mockLocation_withoutMovieId);
    const wrapper = setup();
    expect(wrapper.containsMatchingElement(<Footer />)).toEqual(true);
  });
});
