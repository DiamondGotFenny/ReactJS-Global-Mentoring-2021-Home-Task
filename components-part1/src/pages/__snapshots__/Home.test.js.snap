// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders MovieDetail correctly with movieId query 1`] = `
<styled.div
  className="homePage-Wrapper"
>
  <MovieDetail
    movieId="1"
  />
  <Memo(Content) />
  <Footer />
</styled.div>
`;

exports[`Home renders Search correctly when url without movieId query 1`] = `
<styled.div
  className="homePage-Wrapper"
>
  <Memo(Search) />
  <Memo(Content) />
  <Footer />
</styled.div>
`;
